README.md
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](LICENSE)

## 🧐 What's this?

This `fileee-backup-downloader` is a hastily hacked script that logs into [Fileee](https://www.fileee.com/) with
[pupperteer](https://developers.google.com/web/tools/puppeteer/) and downloads a full backup archive. I use the script
to create my own backup of my documents at regular intervals. Just to be sure.

<br />

## 🎉 Features
- Login to [Fileee](https://www.fileee.com/) using username and password (no 2-Factor supported)
- Download ZIP archive in a temporary folder
- Move archive to given destination

<br />

## 📦 Installation

	npm i -g @sebbo2002/fileee-backup-downloader
    fileee-backup-downloader

    docker run --rm \
        -e FILEEE_USERNAME=**** \
        -e FILEEE_PASSWORD=**** \
        -e BACKUP_DESTINATION=/backup/fileee-backup.zip \
        -v $(pwd):/backup
        @sebbo2002/fileee-backup-downloader

<br />

## 🔧 Usage

| Environment Variable | Description                                                                  |
|:-------------------- |:---------------------------------------------------------------------------- |
|FILEEE_USERNAME       | Username to use to login into Fileee web app                                 |
|FILEEE_PASSWORD       | Password to use to login                                                     |
|FILEEE_BETA           | Use the beta environment instead of production                               |
|BACKUP_DESTINATION    | Full destination path (folder + name), uses `~/fileee-backup.zip` if not set |

<br />

## 📱 Screenshot
![Screenshot of a terminal running the script](https://d.sebbo.net/screenshot-3j3cfHIb10luztGF1Llflg5qgOIVDCg9BNk5GdKo1dK98DaIaYGPex8YLotZeyAaIvmZLc7qeVzPXecIejkcuczcLY0Bvsaioq98.png)

<br />

## 👩‍⚖️️ Legal Stuff

- This tool was created out of private interest and is not affiliated with Fileee or the company behind it
- Licensed under the [MIT license](LICENSE)


CHANGELOG.md
## [3.0.7-develop.3](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.7-develop.2...v3.0.7-develop.3) (2023-01-19)

## [3.0.7-develop.2](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.7-develop.1...v3.0.7-develop.2) (2023-01-12)


### Bug Fixes

* Update password input ([0679ef4](https://github.com/sebbo2002/fileee-backup-downloader/commit/0679ef4eae75531e7e8d028b294bd5e270c5667b))

## [3.0.7-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.6...v3.0.7-develop.1) (2023-01-02)

## [3.0.6](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.5...v3.0.6) (2022-12-27)

## [3.0.6-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.5...v3.0.6-develop.1) (2022-12-27)

## [3.0.5](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.4...v3.0.5) (2022-12-11)

## [3.0.5-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.4...v3.0.5-develop.1) (2022-12-11)

## [3.0.4](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.3...v3.0.4) (2022-10-14)

## [3.0.4-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.3...v3.0.4-develop.1) (2022-10-13)

## [3.0.3](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.2...v3.0.3) (2022-09-18)

## [3.0.3-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.2...v3.0.3-develop.1) (2022-09-16)

## [3.0.2](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.1...v3.0.2) (2022-09-01)

## [3.0.2-develop.2](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.2-develop.1...v3.0.2-develop.2) (2022-08-31)

## [3.0.2-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.1...v3.0.2-develop.1) (2022-08-26)

## [3.0.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.0...v3.0.1) (2022-08-16)

## [3.0.1-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.0...v3.0.1-develop.1) (2022-08-16)

# [3.0.0](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.7...v3.0.0) (2022-07-25)


### Bug Fixes

* Fix "TypeError: page._client.send is not a function" ([3c07a51](https://github.com/sebbo2002/fileee-backup-downloader/commit/3c07a51b991bba62b8cb2f55b16de17bbfaf43b3))


### Build System

* Native ESM support ([7b86a4f](https://github.com/sebbo2002/fileee-backup-downloader/commit/7b86a4f1187c387a3a5792e1fb72d822b04e3631))


### BREAKING CHANGES

* Only Support for node.js ^12.20.0 || >=14.13.1

# [3.0.0-develop.4](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.0-develop.3...v3.0.0-develop.4) (2022-07-25)

# [3.0.0-develop.3](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.0-develop.2...v3.0.0-develop.3) (2022-07-20)

# [3.0.0-develop.2](https://github.com/sebbo2002/fileee-backup-downloader/compare/v3.0.0-develop.1...v3.0.0-develop.2) (2022-07-16)

# [3.0.0-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.8-develop.3...v3.0.0-develop.1) (2022-07-16)


### Build System

* Native ESM support ([7b86a4f](https://github.com/sebbo2002/fileee-backup-downloader/commit/7b86a4f1187c387a3a5792e1fb72d822b04e3631))


### BREAKING CHANGES

* Only Support for node.js ^12.20.0 || >=14.13.1

## [2.0.8-develop.3](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.8-develop.2...v2.0.8-develop.3) (2022-07-13)

## [2.0.8-develop.2](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.8-develop.1...v2.0.8-develop.2) (2022-07-08)

## [2.0.8-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.7...v2.0.8-develop.1) (2022-06-28)


### Bug Fixes

* Fix "TypeError: page._client.send is not a function" ([3c07a51](https://github.com/sebbo2002/fileee-backup-downloader/commit/3c07a51b991bba62b8cb2f55b16de17bbfaf43b3))

## [2.0.7](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.6...v2.0.7) (2022-06-25)

## [2.0.7-develop.4](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.7-develop.3...v2.0.7-develop.4) (2022-06-24)

## [2.0.7-develop.3](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.7-develop.2...v2.0.7-develop.3) (2022-06-22)

## [2.0.7-develop.2](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.7-develop.1...v2.0.7-develop.2) (2022-06-14)

## [2.0.7-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.6...v2.0.7-develop.1) (2022-06-14)

## [2.0.6](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.5...v2.0.6) (2022-06-01)


### Reverts

* Revert "ci: Remove docker setup" ([14a5226](https://github.com/sebbo2002/fileee-backup-downloader/commit/14a5226c828f173fe23b1d0033f825c73e848038))
* Revert "ci: Remove docker setup" ([655068b](https://github.com/sebbo2002/fileee-backup-downloader/commit/655068b3b9c6139181ae87421db5f8144fae3e18))

## [2.0.6-develop.3](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.6-develop.2...v2.0.6-develop.3) (2022-05-26)

## [2.0.6-develop.2](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.6-develop.1...v2.0.6-develop.2) (2022-05-20)


### Reverts

* Revert "ci: Remove GH_TOKEN and use GITHUB_TOKEN" ([b5c2eb6](https://github.com/sebbo2002/fileee-backup-downloader/commit/b5c2eb66170b38bda1e49ad5bb5cf02bd13eb8e4))
* Revert "ci: Remove docker setup" ([655068b](https://github.com/sebbo2002/fileee-backup-downloader/commit/655068b3b9c6139181ae87421db5f8144fae3e18))

## [2.0.6-develop.1](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.5...v2.0.6-develop.1) (2022-05-18)

## [2.0.5](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.4...v2.0.5) (2022-05-14)

## [2.0.5-develop.5](https://github.com/sebbo2002/fileee-backup-downloader/compare/v2.0.5-develop.4...v2.0.5-develop.5) (2022-05-14)


test/lib/index.js
'use strict';

import puppeteer from 'puppeteer';

describe('General', function () {
    this.timeout(60000);

    it('pupperteer should be able to start', async function () {
        const browser = await puppeteer.launch();
        await browser.newPage();

        await browser.close();
    });
});


fileee-backup-downloader.js
#!/usr/bin/env node
'use strict';

const os = require('os');
const path = require('path');
const {mkdtemp, readdir, rename} = require('fs/promises');

const puppeteer = require('puppeteer');
const readline = require('readline');

class FileeeBackupDownloader {
    static async run() {
        console.log('👋🏼 Okay, hi there.');
        console.log('   Let\' do some backups!');
        console.log('');

        const username = process.env.FILEEE_USERNAME;
        if(!username) {
            throw new Error('Unable to run: FILEEE_USERNAME not set.');
        }

        const password = process.env.FILEEE_PASSWORD;
        if(!password) {
            throw new Error('Unable to run: FILEEE_PASSWORD not set.');
        }

        this.logJobStart('⚙️', 'Launch virtual browser');
        const destination = process.env.BACKUP_DESTINATION || path.resolve(process.env.HOME, 'fileee-backup.zip');
        const browser = await puppeteer.launch();
        const page = await browser.newPage();

        try {
            await this.execute({browser, page, username, password, destination});
        }
        catch(error) {
            this.logJobEnd();

            try {
                await page.screenshot({path: new Date().getTime() + '-error.png'});
                await browser.close();
            }
            catch(error) {
                console.log(error);
            }

            throw error;
        }
    }

    static async execute({browser, page, username, password, destination}) {
        this.logJobStart('🌍️', 'Open fileee web app');
        await page.goto('https://my.fileee.com/account');

        this.logJobStart('👤', 'Enter username');
        const $usernameInput = await page.waitForSelector('[name="username"]');
        await $usernameInput.type(username);
        await $usernameInput.press('Enter');

        this.logJobStart('🔑️', 'Enter password');
        const $passwordInput = await page.waitForSelector('[name="password"]');
        await $passwordInput.type(password);
        await $passwordInput.press('Enter');
        await page.waitForTimeout(2500);

        this.logJobStart('📑', 'Open download layer');
        const $downloadButton = await page.waitForSelector('.grid-noGutter-spaceBetween button');
        await page.waitForTimeout(1000);
        await $downloadButton.click();
        await page.waitForTimeout(100);

        await Promise.race([
            page.waitForSelector('.ReactModalPortal input[type="password"]'),
            page.waitForSelector('.ReactModalPortal .mdc-typography--caption span')
        ]);

        const $confirmPasswordInput = await page.$('.ReactModalPortal input[type="password"]');
        if($confirmPasswordInput) {
            this.logJobStart('🔑', 'Enter password (again)');
            await $confirmPasswordInput.type(password);
            await $confirmPasswordInput.press('Enter');
        }

        this.logJobStart('🔄', 'Prepare download');
        const tmpDir = await mkdtemp(path.join(os.tmpdir(), 'fileee-download-'));
        await page._client.send('Page.setDownloadBehavior', {
            behavior: 'allow',
            downloadPath: tmpDir,
        });

        await page.waitForSelector('.ReactModalPortal .mdc-typography--caption span');
        const interval = setInterval(async () => {
            page.$eval('.ReactModalPortal .mdc-typography--caption span', s => s.textContent)
                .then(state => {
                    if(!process.env.FILEEE_NO_LOG) {
                        process.stdout.write(`\r🔄 Prepare download (${state.split(' ')[0]})`);
                    }
                })
                .catch(error => {/* ignore errors here */});
        }, 1000);
        await page.waitForResponse(
            r => r.url().startsWith('https://my.fileee.com/api/v1/zip/download/'),
            {timeout: 30 * 60 * 1000}
        );
        clearInterval(interval);

        this.logJobStart('⬇️', 'Download archive');
        let filePath = null;
        for(const start = new Date().getTime(); new Date().getTime() - start < 30 * 60 * 1000; ) {
            const files = await readdir(tmpDir);
            const zipFileName = files.find(file => file.endsWith('.zip'));
            if(zipFileName) {
                filePath = path.join(tmpDir, zipFileName);
                break;
            }

            await page.waitForTimeout(1000);
        }

        this.logJobStart('✋🏼', 'Close browser');
        await browser.close();

        this.logJobStart('⏩', 'Move file to destination');
        await rename(filePath, destination);
        this.logJobEnd();

        if(!process.env.FILEEE_NO_LOG) {
            console.log('\n🎉 Completed');
            console.log(`   Backup path: ${destination}`);
        }
    }

    static logJobStart(emoji, name) {
        if(process.env.FILEEE_NO_LOG) {
            return;
        }
        if(this.logJobStart.current) {
            this.logJobEnd();
        }

        this.logJobStart.current = [name, new Date().getTime()];
        process.stdout.write(`${emoji} ${name}`);
    }

    static logJobEnd() {
        if(!this.logJobStart.current || process.env.FILEEE_NO_LOG) {
            return;
        }

        const [name, started] = this.logJobStart.current;
        const duration = new Date().getTime() - started;

        this.logJobStart.current = null;
        let durationTxt = '';

        if(duration >= 94000) {
            durationTxt = ` [${Math.floor(duration / 1000 / 60)}:${Math.ceil((duration / 1000) % 60).toString().padStart(2, '0')}]`;
        }
        else if(duration > 1000) {
            durationTxt = ` [${Math.ceil(duration / 1000)}s]`;
        }
        else if(duration > 50) {
            durationTxt = ` [${duration}ms]`;
        }

        readline.clearLine(process.stdout);
        readline.cursorTo(process.stdout, 0);
        console.log(`✅  ${name}${durationTxt}`);
    }
}

module.exports = FileeeBackupDownloader;


src/lib/index.js
#!/usr/bin/env node
'use strict';

import {tmpdir} from 'os';
import readline from 'readline';
import {resolve, join} from 'path';
import {promises as fsPromises} from 'fs';
import puppeteer from 'puppeteer';

const {mkdtemp, readdir, rename} = fsPromises;

export default class FileeeBackupDownloader {
    static async run() {
        if(!process.env.FILEEE_NO_LOG) {
            console.log('👋🏼 Okay, hi there.');
            console.log('   Let\' do some backups!');
            console.log('');
        }

        const username = process.env.FILEEE_USERNAME;
        if(!username) {
            throw new Error('Unable to run: FILEEE_USERNAME not set.');
        }

        const password = process.env.FILEEE_PASSWORD;
        if(!password) {
            throw new Error('Unable to run: FILEEE_PASSWORD not set.');
        }

        this.logJobStart('⚙️', 'Launch virtual browser');
        const destination = process.env.BACKUP_DESTINATION || resolve(process.env.HOME, 'fileee-backup.zip');
        const browser = await puppeteer.launch();
        const page = await browser.newPage();

        try {
            await this.execute({browser, page, username, password, destination});
        }
        catch(error) {
            this.logJobEnd();

            try {
                await page.screenshot({path: new Date().getTime() + '-error.png'});
                await browser.close();
            }
            catch(error) {
                console.log(error);
            }

            throw error;
        }
    }

    static async execute({browser, page, username, password, destination}) {
        this.logJobStart('🌍️', 'Open fileee web app');
        await page.goto('https://' + (process.env.FILEEE_BETA ? 'beta' : 'my') + '.fileee.com/account/general/statistics');

        this.logJobStart('👤', 'Enter username');
        const $usernameInput = await page.waitForSelector('[name="username"]');
        await $usernameInput.type(username);
        await $usernameInput.press('Enter');

        this.logJobStart('🔑️', 'Enter password');
        const $passwordInput = await page.waitForSelector('[type="password"]');
        await $passwordInput.type(password);
        await $passwordInput.press('Enter');
        await page.waitForTimeout(2500);

        this.logJobStart('📑', 'Open download layer');
        const $downloadButton = await page.waitForSelector('button');
        await page.waitForTimeout(1000);
        await $downloadButton.click();
        await page.waitForTimeout(100);

        await Promise.race([
            page.waitForSelector('.ReactModalPortal input[type="password"]'),
            page.waitForSelector('.ReactModalPortal .mdc-typography--caption span')
        ]);

        let $confirmPasswordInput = null;
        try {
            $confirmPasswordInput = await page.waitForSelector('.ReactModalPortal input[type="password"]', {
                timeout: 5000
            });
        }
        catch(error) {
            // ignore error
        }
        if($confirmPasswordInput) {
            this.logJobStart('🔑', 'Enter password (again)');
            await $confirmPasswordInput.type(password);
            await $confirmPasswordInput.press('Enter');
        }

        this.logJobStart('🔄', 'Prepare download');
        const tmpDir = await mkdtemp(join(tmpdir(), 'fileee-download-'));
        await page._client().send('Page.setDownloadBehavior', {
            behavior: 'allow',
            downloadPath: tmpDir,
        });

        await page.waitForSelector('.ReactModalPortal .mdc-typography--caption span');
        const interval = setInterval(async () => {
            page.$eval('.ReactModalPortal .mdc-typography--caption span', s => s.textContent)
                .then(state => {
                    if(!process.env.FILEEE_NO_LOG) {
                        process.stdout.write(`\r🔄 Prepare download (${state.split(' ')[0]})`);
                    }
                })
                .catch(() => {/* ignore errors here */});
        }, 1000);
        await page.waitForResponse(
            r => {
                const isDownload = r.url().includes('/api/v1/zip/download/');
                if(isDownload && r.status() !== 200) {
                    throw new Error(
                        `Unable to download backup from ${r.url()}: ` +
                        `Server responded with code ${r.status()} ${r.statusText()}`
                    );
                }

                return isDownload;
            },
            {timeout: 30 * 60 * 1000}
        );
        clearInterval(interval);

        this.logJobStart('⬇️', 'Download archive');
        let filePath = null;
        for(const start = new Date().getTime(); new Date().getTime() - start < 30 * 60 * 1000; ) {
            const files = await readdir(tmpDir);
            const zipFileName = files.find(file => file.endsWith('.zip'));
            if(zipFileName) {
                filePath = join(tmpDir, zipFileName);
                break;
            }

            await page.waitForTimeout(1000);
        }

        this.logJobStart('✋🏼', 'Close browser');
        await browser.close();

        this.logJobStart('⏩', 'Move file to destination');
        await rename(filePath, destination);
        this.logJobEnd();

        if(!process.env.FILEEE_NO_LOG) {
            console.log('\n🎉 Completed');
            console.log(`   Backup path: ${destination}`);
        }
    }

    static logJobStart(emoji, name) {
        if(process.env.FILEEE_NO_LOG) {
            return;
        }
        if(this.logJobStart.current) {
            this.logJobEnd();
        }

        this.logJobStart.current = [name, new Date().getTime()];
        process.stdout.write(`${emoji} ${name}`);
    }

    static logJobEnd() {
        if(!this.logJobStart.current || process.env.FILEEE_NO_LOG) {
            return;
        }

        const [name, started] = this.logJobStart.current;
        const duration = new Date().getTime() - started;

        this.logJobStart.current = null;
        let durationTxt = '';

        if(duration >= 94000) {
            durationTxt = ` [${Math.floor(duration / 1000 / 60)}:${Math.ceil((duration / 1000) % 60).toString().padStart(2, '0')}]`;
        }
        else if(duration > 1000) {
            durationTxt = ` [${Math.ceil(duration / 1000)}s]`;
        }
        else if(duration > 50) {
            durationTxt = ` [${duration}ms]`;
        }

        readline.clearLine(process.stdout);
        readline.cursorTo(process.stdout, 0);
        console.log(`✅  ${name}${durationTxt}`);
    }
}
